File: TopInterview150/Backtracking/Combinations/solution.py
class Solution(object):
    
    def backtrack(self, result, comb, n, k, start):
        if len(comb) == k:
            result.append(list(comb))
            return
        
        for i in range(start, n+1):
            comb.append(i)
            self.backtrack(result, comb, n, k, i+1)
            comb.pop()
        
    def combine(self, n, k):
        """
        :type n: int
        :type k: int
        :rtype: List[List[int]]
        """
        
        result = []
        comb = []

        self.backtrack(result, comb, n, k, 1)
        return result

File: TopInterview150/Backtracking/Combinations/solution.cpp
class Solution {
    public:
        vector<vector<int>> combine(int n, int k) {
            vector<vector<int>> result;
            vector<int> comb;

            backtrack(result, comb, n, k, 1);
            return result;
        }

        void backtrack(
            vector<vector<int>>& result,
            vector<int>& comb, 
            //pass by reference &s go in the function definition, not in the invocation
            int n, 
            int k, 
            int start
        )
        {
            if (comb.size() == k)
            {
                result.push_back(comb);
                return;
            }

            for (int i = start; i <= n; i++)
            {
                comb.push_back(i);
                backtrack(result, comb, n, k, i+1);
                comb.pop_back();
            }
        }
    };